service/controller/v18/resource/cloudformation/current.go

```diff
@@ -23,11 +23,43 @@ func (r *Resource) GetCurrentState(ctx context.Context, obj interface{}) (interf
 
 	stackName := key.MainGuestStackName(customObject)
 
-	sc, err := controllercontext.FromContext(ctx)
+	ctlCtx, err := controllercontext.FromContext(ctx)
 	if err != nil {
 		return StackState{}, microerror.Mask(err)
 	}
 
+	if key.IsDeleted(customObject) {
+		stackNames := []string{
+			key.MainGuestStackName(customObject),
+			key.MainHostPreStackName(customObject),
+			key.MainHostPostStackName(customObject),
+		}
+
+		for _, stackName := range stackNames {
+			r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("finding stack %#q in the AWS API", stackName))
+
+			in := &cloudformation.DescribeStacksInput{
+				StackName: aws.String(stackName),
+			}
+
+			out, err := ctlCtx.AWSClient.CloudFormation.DescribeStacks(in)
+			if cloudformationservice.IsStackNotFound(err) {
+				// This handling is far from perfect. We use different
+				// packages here. This is all going to be addressed in
+				// scope of
+				// https://github.com/giantswarm/giantswarm/issues/3783.
```

Comment:
We have the handling of stack statuses in several places and should maybe have some common library and understanding of how to work with these states. Like having a state machine with hooks or something. Is the issue referenced here where we want to address this or is this a different issue?

Reply:
I'll address it here. I even have some prototypes written.